File: apps/guardrail-server/src/guardrail/core/circuit_breaker.py
"""Circuit breaker implementation for fault-tolerant model calls.

Implements the circuit breaker pattern to prevent cascade failures:
- CLOSED: Normal operation, requests pass through
- OPEN: After N failures, all requests fail immediately
- HALF_OPEN: After recovery timeout, allow probe requests

Reference: https://martinfowler.com/bliki/CircuitBreaker.html
"""

import asyncio
import time
from enum import Enum
from dataclasses import dataclass, field
from prometheus_client import Gauge

from guardrail.config import settings


class CircuitState(Enum):
    """Circuit breaker states."""
    CLOSED = 0  # Normal operation
    OPEN = 1    # Rejecting all requests
    HALF_OPEN = 2  # Testing if service recovered


# Prometheus metric for circuit breaker state
CIRCUIT_STATE_GAUGE = Gauge(
    "guardrail_circuit_breaker_state",
    "Circuit breaker state (0=closed, 1=open, 2=half_open)",
    ["model_name"],
)


@dataclass
class CircuitBreaker:
    """Circuit breaker for a single model service.
    
    Usage:
        cb = CircuitBreaker("prompt-guard")
        
        if not cb.allow_request():
            raise CircuitOpenError()
        
        try:
            result = await call_model()
            cb.record_success()
        except Exception:
            cb.record_failure()
    """
    
    name: str
    failure_threshold: int = field(default_factory=lambda: settings.CB_FAILURE_THRESHOLD)
    recovery_timeout: float = field(default_factory=lambda: settings.CB_RECOVERY_TIMEOUT)
    success_threshold: int = field(default_factory=lambda: settings.CB_SUCCESS_THRESHOLD)
    
    # Internal state
    _state: CircuitState = field(default=CircuitState.CLOSED, init=False)
    _failure_count: int = field(default=0, init=False)
    _success_count: int = field(default=0, init=False)
    _last_failure_time: float = field(default=0.0, init=False)
    _lock: asyncio.Lock = field(default_factory=asyncio.Lock, init=False)
    
    def __post_init__(self):
        """Initialize metrics after dataclass creation."""
        CIRCUIT_STATE_GAUGE.labels(model_name=self.name).set(self._state.value)
    
    @property
    def state(self) -> CircuitState:
        """Get current circuit state, checking for timeout recovery."""
        if self._state == CircuitState.OPEN:
            # Check if recovery timeout has passed
            if time.monotonic() - self._last_failure_time >= self.recovery_timeout:
                self._transition_to(CircuitState.HALF_OPEN)
        return self._state
    
    def _transition_to(self, new_state: CircuitState) -> None:
        """Transition to a new state and update metrics."""
        old_state = self._state
        self._state = new_state
        CIRCUIT_STATE_GAUGE.labels(model_name=self.name).set(new_state.value)
        
        # Reset counters on state change
        if new_state == CircuitState.CLOSED:
            self._failure_count = 0
        elif new_state == CircuitState.HALF_OPEN:
            self._success_count = 0
    
    def allow_request(self) -> bool:
        """Check if a request should be allowed.
        
        Returns:
            True if request can proceed, False if circuit is open
        """
        current_state = self.state  # Triggers timeout check
        
        if current_state == CircuitState.CLOSED:
            return True
        elif current_state == CircuitState.HALF_OPEN:
            # Allow probe requests in half-open state
            return True
        else:  # OPEN
            return False
    
    async def record_success(self) -> None:
        """Record a successful request."""
        async with self._lock:
            if self._state == CircuitState.HALF_OPEN:
                self._success_count += 1
                if self._success_count >= self.success_threshold:
                    self._transition_to(CircuitState.CLOSED)
            elif self._state == CircuitState.CLOSED:
                # Reset failure count on success
                self._failure_count = 0
    
    async def record_failure(self) -> None:
        """Record a failed request."""
        async with self._lock:
            self._failure_count += 1
            self._last_failure_time = time.monotonic()
            
            if self._state == CircuitState.HALF_OPEN:
                # Any failure in half-open goes back to open
                self._transition_to(CircuitState.OPEN)
            elif self._state == CircuitState.CLOSED:
                if self._failure_count >= self.failure_threshold:
                    self._transition_to(CircuitState.OPEN)
    
    def force_close(self) -> None:
        """Force the circuit to close (for debugging/recovery)."""
        self._transition_to(CircuitState.CLOSED)
    
    def force_open(self) -> None:
        """Force the circuit to open (for testing)."""
        self._last_failure_time = time.monotonic()
        self._transition_to(CircuitState.OPEN)
    
    def get_status(self) -> dict:
        """Get current circuit breaker status."""
        return {
            "name": self.name,
            "state": self.state.name,
            "failure_count": self._failure_count,
            "success_count": self._success_count,
            "last_failure_time": self._last_failure_time,
        }


class CircuitOpenError(Exception):
    """Raised when circuit breaker is open."""
    
    def __init__(self, model_name: str):
        self.model_name = model_name
        super().__init__(f"Circuit breaker open for model: {model_name}")


# Global circuit breakers for each model
_circuit_breakers: dict[str, CircuitBreaker] = {}


def get_circuit_breaker(model_name: str) -> CircuitBreaker:
    """Get or create a circuit breaker for a model."""
    if model_name not in _circuit_breakers:
        _circuit_breakers[model_name] = CircuitBreaker(name=model_name)
    return _circuit_breakers[model_name]


def get_all_circuit_breakers() -> dict[str, CircuitBreaker]:
    """Get all circuit breakers (for debugging)."""
    return _circuit_breakers

File: apps/guardrail-server-go/internal/circuitbreaker/circuitbreaker.go
// Package circuitbreaker implements the circuit breaker pattern for fault-tolerant model calls.
//
// Implements the circuit breaker pattern to prevent cascade failures:
//   - CLOSED: Normal operation, requests pass through
//   - OPEN: After N failures, all requests fail immediately
//   - HALF_OPEN: After recovery timeout, allow probe requests
//
// Reference: https://martinfowler.com/bliki/CircuitBreaker.html
package circuitbreaker

import (
	"sync"
	"time"

	gocommon "github.com/playground/packages/go-common"
	"github.com/prometheus/client_golang/prometheus"
)

// State represents the circuit breaker state.
type State int

const (
	// StateClosed is normal operation
	StateClosed State = 0
	// StateOpen is rejecting all requests
	StateOpen State = 1
	// StateHalfOpen is testing if service recovered
	StateHalfOpen State = 2
)

// String returns the state name.
func (s State) String() string {
	switch s {
	case StateClosed:
		return "CLOSED"
	case StateOpen:
		return "OPEN"
	case StateHalfOpen:
		return "HALF_OPEN"
	default:
		return "UNKNOWN"
	}
}

// CircuitBreaker provides circuit breaker functionality for a single model service.
type CircuitBreaker struct {
	name             string
	failureThreshold int
	recoveryTimeout  time.Duration
	successThreshold int

	mu              sync.Mutex
	state           State
	failureCount    int
	successCount    int
	lastFailureTime time.Time

	// Prometheus gauge for state
	stateGauge *prometheus.GaugeVec
}

// New creates a new circuit breaker.
func New(name string, failureThreshold, successThreshold int, recoveryTimeout time.Duration, stateGauge *prometheus.GaugeVec) *CircuitBreaker {
	cb := &CircuitBreaker{
		name:             name,
		failureThreshold: failureThreshold,
		successThreshold: successThreshold,
		recoveryTimeout:  recoveryTimeout,
		state:            StateClosed,
		stateGauge:       stateGauge,
	}

	// Initialize metric
	if stateGauge != nil {
		stateGauge.WithLabelValues(name).Set(float64(StateClosed))
	}

	return cb
}

// AllowRequest checks if a request should be allowed.
func (cb *CircuitBreaker) AllowRequest() bool {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	// Check for recovery timeout
	if cb.state == StateOpen {
		if time.Since(cb.lastFailureTime) >= cb.recoveryTimeout {
			cb.transitionTo(StateHalfOpen)
		}
	}

	switch cb.state {
	case StateClosed, StateHalfOpen:
		return true
	default:
		return false
	}
}

// RecordSuccess records a successful request.
func (cb *CircuitBreaker) RecordSuccess() {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	switch cb.state {
	case StateHalfOpen:
		cb.successCount++
		if cb.successCount >= cb.successThreshold {
			cb.transitionTo(StateClosed)
		}
	case StateClosed:
		cb.failureCount = 0
	}
}

// RecordFailure records a failed request.
func (cb *CircuitBreaker) RecordFailure() {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	cb.failureCount++
	cb.lastFailureTime = time.Now()

	switch cb.state {
	case StateHalfOpen:
		cb.transitionTo(StateOpen)
	case StateClosed:
		if cb.failureCount >= cb.failureThreshold {
			cb.transitionTo(StateOpen)
		}
	}
}

// transitionTo transitions to a new state. Must be called with lock held.
func (cb *CircuitBreaker) transitionTo(newState State) {
	cb.state = newState

	if cb.stateGauge != nil {
		cb.stateGauge.WithLabelValues(cb.name).Set(float64(newState))
	}

	// Reset counters on state change
	switch newState {
	case StateClosed:
		cb.failureCount = 0
	case StateHalfOpen:
		cb.successCount = 0
	}
}

// ForceClose forces the circuit to close.
func (cb *CircuitBreaker) ForceClose() {
	cb.mu.Lock()
	defer cb.mu.Unlock()
	cb.transitionTo(StateClosed)
}

// ForceOpen forces the circuit to open.
func (cb *CircuitBreaker) ForceOpen() {
	cb.mu.Lock()
	defer cb.mu.Unlock()
	cb.lastFailureTime = time.Now()
	cb.transitionTo(StateOpen)
}

// GetStatus returns the current status.
func (cb *CircuitBreaker) GetStatus() gocommon.CircuitBreakerStatus {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	// Check for recovery timeout
	currentState := cb.state
	if cb.state == StateOpen && time.Since(cb.lastFailureTime) >= cb.recoveryTimeout {
		currentState = StateHalfOpen
	}

	return gocommon.CircuitBreakerStatus{
		Name:            cb.name,
		State:           currentState.String(),
		FailureCount:    cb.failureCount,
		SuccessCount:    cb.successCount,
		LastFailureTime: float64(cb.lastFailureTime.Unix()),
	}
}

// State returns the current state.
func (cb *CircuitBreaker) State() State {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	// Check for recovery timeout
	if cb.state == StateOpen && time.Since(cb.lastFailureTime) >= cb.recoveryTimeout {
		cb.transitionTo(StateHalfOpen)
	}

	return cb.state
}

// Registry holds all circuit breakers.
type Registry struct {
	breakers map[string]*CircuitBreaker
	mu       sync.RWMutex

	// Config for new breakers
	failureThreshold int
	successThreshold int
	recoveryTimeout  time.Duration
	stateGauge       *prometheus.GaugeVec
}

// NewRegistry creates a new circuit breaker registry.
func NewRegistry(failureThreshold, successThreshold int, recoveryTimeout time.Duration, stateGauge *prometheus.GaugeVec) *Registry {
	return &Registry{
		breakers:         make(map[string]*CircuitBreaker),
		failureThreshold: failureThreshold,
		successThreshold: successThreshold,
		recoveryTimeout:  recoveryTimeout,
		stateGauge:       stateGauge,
	}
}

// Get returns the circuit breaker for a model, creating it if necessary.
func (r *Registry) Get(name string) *CircuitBreaker {
	r.mu.RLock()
	cb, exists := r.breakers[name]
	r.mu.RUnlock()

	if exists {
		return cb
	}

	r.mu.Lock()
	defer r.mu.Unlock()

	// Double-check after acquiring write lock
	if cb, exists = r.breakers[name]; exists {
		return cb
	}

	cb = New(name, r.failureThreshold, r.successThreshold, r.recoveryTimeout, r.stateGauge)
	r.breakers[name] = cb
	return cb
}

// GetAll returns all circuit breakers.
func (r *Registry) GetAll() map[string]*CircuitBreaker {
	r.mu.RLock()
	defer r.mu.RUnlock()

	result := make(map[string]*CircuitBreaker, len(r.breakers))
	for k, v := range r.breakers {
		result[k] = v
	}
	return result
}
